tests/assets/packages/agent_service_prototype.py

```diff
@@ -0,0 +1,290 @@
+    human_description = "ToolThatAlwaysResponseSynchronouslyAndStatically"
+
+    def run(self, tool_input: List[Block], context: AgentContext) -> Union[List[Block], Task[Any]]:
+        return [Block(text="This was Synchronous and Static")]
+
+
+class ToolThatAlwaysResponseSynchronouslyAndDynamically(Tool):
+    name = "ToolThatAlwaysResponseSynchronouslyAndDynamically"
+    ai_description = "ToolThatAlwaysResponseSynchronouslyAndDynamically"
+    human_description = "ToolThatAlwaysResponseSynchronouslyAndDynamically"
+
+    def run(self, tool_input: List[Block], context: AgentContext) -> Union[List[Block], Task[Any]]:
+        ret = []
+        for block in tool_input:
+            ret.append(
+                Block(text=f"This was synchronous and Dynamic. Input block was: {block.text}")
+            )
+        return ret
+
+
+class ToolThatDynamicallyPlansMoreAsyncWork(Tool):
+    name = "ToolThatDynamicallyPlansMoreAsyncWork"
+    ai_description = "ToolThatDynamicallyPlansMoreAsyncWork"
+    human_description = "ToolThatDynamicallyPlansMoreAsyncWork"
+
+    def run(self, tool_input: List[Block], context: AgentContext) -> Union[List[Block], Task[Any]]:
+        # First let's call the static async tool!
+        dyamic_plan_step_1 = ToolAction(
+            tool_name="ToolThatAlwaysResponseAsynchronouslyAndStatically", input=ToolData()
+        )
+
+        # This ToolAction represents the eventual completion of this step.
+        dyamic_plan_step_1_output = context.invoke_tool(dyamic_plan_step_1)
+        logging.info(
+            f"{self.name}: Output ToolData of Tool1 Invoke was {dyamic_plan_step_1_output.output}"
+        )
+
+        # Now let's call the dynamic async tool... providing it the input of the prior tool!
+        # We don't need to do any manipulation of input since the broader system guarantees
+        # everything is [Block] -> [Block].
+
+        # This is the key magic right here. We can just set step2.input = step1.output and it works.. even
+        # thought none of the actual processing has been done, or data created.
+
+        # Note how the ToolData object also has a carve-out for a file_id. This isn't yet pursued in the unit
+        # tests on this branch but added to illustrate the map-reduce scenario: in which the initiator of the map
+        # operation synchronously (1) creates a file. and (2) spawns TONS of tasks. then (3) writes those task IDs
+        # to the file. THEN it schedules a future ToolAction that depends on those tasks, and takes as input the ToolData
+        # defined by that file's ID. That means that next ToolAction (1) only happens when the map operations are done, and
+        # (2) is passed, automatically, as its input, a set of blocks containing the UUIDs of the tasks with the results.
+        dymamic_plan_step_2 = ToolAction(
+            tool_name="ToolThatAlwaysResponseAsynchronouslyAndDynamically",
+            input=dyamic_plan_step_1_output.output,
```

Comment:
this style _appears_ to assume the following:
- the entirety of a `Tool`s input in contained in the output of preceding tool.
- we must know the mapping of inputs -> outputs between `Tools` up front.

Is that correct?

If so, how would we work around those limitations?

Context: the advantage of a _distinct_ planner is that it can use the whole execution context to decide on what the input for the next tool in the chain should be.